src/Compilers/Core/Portable/SourceGeneration/Nodes/IncrementalValueSources.cs

```diff
@@ -31,102 +28,21 @@ internal IncrementalValueSources(PerGeneratorInputNodes.Builder perGeneratorBuil
-
-        //only used for back compat in the adaptor
-        internal IncrementalValueSource<ISyntaxContextReceiver> CreateSyntaxReceiver() => new IncrementalValueSource<ISyntaxContextReceiver>(_perGeneratorBuilder.GetOrCreateReceiverNode());
     }
 
     /// <summary>
     /// Holds input nodes that are shared between generators and always exist
     /// </summary>
     internal static class SharedInputNodes
     {
-        public static readonly InputNode<Compilation> Compilation = new InputNode<Compilation>();
-
-        public static readonly InputNode<ParseOptions> ParseOptions = new InputNode<ParseOptions>();
-
-        public static readonly InputNode<AdditionalText> AdditionalTexts = new InputNode<AdditionalText>();
-
-        public static readonly InputNode<SyntaxTree> SyntaxTrees = new InputNode<SyntaxTree>();
-
-        public static readonly InputNode<AnalyzerConfigOptionsProvider> AnalyzerConfigOptions = new InputNode<AnalyzerConfigOptionsProvider>();
-    }
-
-    /// <summary>
-    /// Holds input nodes that are created per-generator
-    /// </summary>
-    internal sealed class PerGeneratorInputNodes
-    {
-        public static readonly PerGeneratorInputNodes Empty = new PerGeneratorInputNodes(receiverNode: null, ImmutableArray<ISyntaxTransformNode>.Empty);
-
-        private PerGeneratorInputNodes(InputNode<ISyntaxContextReceiver>? receiverNode, ImmutableArray<ISyntaxTransformNode> transformNodes)
-        {
-            this.ReceiverNode = receiverNode;
-            this.TransformNodes = transformNodes;
-        }
-
-        public ImmutableArray<ISyntaxTransformNode> TransformNodes { get; }
-
-        public InputNode<ISyntaxContextReceiver>? ReceiverNode { get; }
-
-        public sealed class Builder
-        {
-            private InputNode<ISyntaxContextReceiver>? _receiverNode;
+        public static readonly InputNode<Compilation> Compilation = new InputNode<Compilation>(b => ImmutableArray.Create(b.Compilation));
 
-            private ArrayBuilder<ISyntaxTransformNode>? _transformNodes;
+        public static readonly InputNode<ParseOptions> ParseOptions = new InputNode<ParseOptions>(b => ImmutableArray.Create(b.DriverState.ParseOptions));
 
-            bool _disposed = false;
+        public static readonly InputNode<AdditionalText> AdditionalTexts = new InputNode<AdditionalText>(b => b.DriverState.AdditionalTexts);
 
-            public InputNode<ISyntaxContextReceiver> GetOrCreateReceiverNode()
-            {
-                Debug.Assert(!_disposed);
-                if (_receiverNode is null)
-                {
-                    // this is called by only internal code which we know to be thread safe
-                    _receiverNode = new InputNode<ISyntaxContextReceiver>();
-                }
-                return _receiverNode;
-            }
+        public static readonly InputNode<SyntaxTree> SyntaxTrees = new InputNode<SyntaxTree>(b => b.Compilation.SyntaxTrees.ToImmutableArray());
```

Comment:
This is one where we may be able to leverage an efficient `IEqualityComparer`. Would two `SytnaxTree` with the same checksum and path be equal and hence cached here? Basically don't think we need object identity. 

Probably a minor issue, mostly help with `PostInit` generated files I imagine. Something to consider though.

Reply:
The post init trees are only created once, so object equality already works for them. I could imagine it being much more efficient if for instance a user adds, then immediately deletes a character. I'm actually unsure on how the IDE actually runs this though. @jasonmalinowski  would we see the intermediate tree for each key press, or is it batched in some way?

---
In reply to: [632882883](https://github.com/dotnet/roslyn/pull/53383#discussion_r632882883) [](ancestors = 632882883)